src/proto/factories/layer_graph_factory.cpp

```diff
@@ -194,17 +150,18 @@ std::vector<std::unique_ptr<Layer>> construct_layer_graph(
     auto proto_datatype = proto_layer.datatype();
 
     // Construct layer
-    std::unique_ptr<Layer> l;
+    OwningLayerPtr l;
 #define TEMPLATE_INSTANTIATION(TensorDataType, T_layout, T_device)      \
     do {                                                                \
       if (proto_datatype == TypeToProtoDataType<TensorDataType>::value  \
           && layout == T_layout                                         \
           && device == T_device) {                                      \
-        l = construct_layer<TensorDataType, T_layout, T_device>(        \
+        auto l_ = construct_layer<TensorDataType, T_layout, T_device>(  \
```

Comment:
`std::shared_ptr<T>` is both assignable and constructible from `std::unique_ptr<T>` rvalues.
```suggestion
        l = construct_layer<TensorDataType, T_layout, T_device>(  \
```